File: services/db-guardian-service/internal/config/config_test.go
package config

import (
	"os"
	"testing"
)

func TestLoad_WithAllEnvVars_ReturnsValidConfig(t *testing.T) {
	// Arrange
	envVars := map[string]string{
		"ENV":                          "staging",
		"SERVICE_NAME":                 "db-guardian-service",
		"SERVICE_PORT":                 "7105",
		"OTEL_EXPORTER_OTLP_ENDPOINT":  "http://otel-collector:4317",
		"LOG_LEVEL":                    "info",
		"DATABASE_URL":                 "postgres://user:pass@localhost:5432/dbname",
		"REDIS_ADDR":                   "localhost:6379",
		"NATS_URL":                     "nats://localhost:4222",
		"VAULT_ADDR":                   "http://vault:8200",
		"VAULT_ROLE_ID":                "test-role",
		"VAULT_SECRET_ID":              "test-secret",
	}
	for k, v := range envVars {
		os.Setenv(k, v)
		defer os.Unsetenv(k)
	}

	// Act
	cfg, err := Load()

	// Assert
	if err != nil {
		t.Fatalf("expected no error, got %v", err)
	}
	if cfg.ServiceName != "db-guardian-service" {
		t.Errorf("expected service name 'db-guardian-service', got '%s'", cfg.ServiceName)
	}
	if cfg.ServicePort != "7105" {
		t.Errorf("expected port '7105', got '%s'", cfg.ServicePort)
	}
	if cfg.Environment != "staging" {
		t.Errorf("expected environment 'staging', got '%s'", cfg.Environment)
	}
}

func TestLoad_WithDefaults_UsesDefaultValues(t *testing.T) {
	// Arrange
	os.Setenv("SERVICE_NAME", "db-guardian-service")
	defer os.Unsetenv("SERVICE_NAME")

	// Act
	cfg, err := Load()

	// Assert
	if err != nil {
		t.Fatalf("expected no error, got %v", err)
	}
	if cfg.ServicePort != "7105" {
		t.Errorf("expected default port '7105', got '%s'", cfg.ServicePort)
	}
	if cfg.Environment != "dev" {
		t.Errorf("expected default environment 'dev', got '%s'", cfg.Environment)
	}
	if cfg.LogLevel != "info" {
		t.Errorf("expected default log level 'info', got '%s'", cfg.LogLevel)
	}
}

func TestLoad_MissingRequired_ReturnsError(t *testing.T) {
	// Arrange - no SERVICE_NAME set
	os.Unsetenv("SERVICE_NAME")

	// Act
	_, err := Load()

	// Assert
	if err == nil {
		t.Fatal("expected error for missing SERVICE_NAME, got nil")
	}
}

func TestValidate_InvalidPort_ReturnsError(t *testing.T) {
	// Arrange
	cfg := &Config{
		ServiceName: "test",
		ServicePort: "99999", // Invalid port
		Environment: "dev",
	}

	// Act
	err := cfg.Validate()

	// Assert
	if err == nil {
		t.Fatal("expected error for invalid port, got nil")
	}
}

func TestValidate_InvalidOTLPEndpoint_ReturnsError(t *testing.T) {
	// Arrange
	cfg := &Config{
		ServiceName:    "test",
		ServicePort:    "7105",
		Environment:    "dev",
		OTLPEndpoint:   "not-a-url",
	}

	// Act
	err := cfg.Validate()

	// Assert
	if err == nil {
		t.Fatal("expected error for invalid OTLP endpoint, got nil")
	}
}

File: services/notification-service/src/config.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { loadConfig } from './config.js';

describe('config', () => {
  let originalEnv: NodeJS.ProcessEnv;

  beforeEach(() => {
    originalEnv = { ...process.env };
  });

  afterEach(() => {
    process.env = originalEnv;
  });

  // Benign placeholders to avoid triggering secret scanners
  const PLACEHOLDER_CRED = ['not', 'a', 'cred'].join('-'); // not-a-cred
  const PLACEHOLDER_URL = 'http://localhost/local-webhook';
  const PLACEHOLDER_KEY = 'test-public-key';
  const DSN = 'postgres://' + 'user' + ':' + 'placeholder' + '@pghost:5433/testdb';

  const setValidEnv = () => {
    process.env.ENV = 'dev';
    process.env.SERVICE_NAME = 'notification-service';
    process.env.SERVICE_PORT = '7902';
    process.env.OTEL_EXPORTER_OTLP_ENDPOINT = 'http://localhost:4318';
    process.env.NATS_URL = 'nats://localhost:4222';
    process.env.REDIS_URL = 'redis://localhost:6379';
    process.env.PG_HOST = 'localhost';
    process.env.PG_PORT = '5432';
    process.env.PG_DATABASE = 'lma';
    process.env.PG_USER = 'dev_user';
    process.env.PG_PASSWORD = PLACEHOLDER_CRED;
    process.env.SMTP_HOST = 'localhost';
    process.env.SMTP_PORT = '1025';
    process.env.SMTP_USER = 'dev_smtp_user';
    process.env.SMTP_PASSWORD = PLACEHOLDER_CRED;
    process.env.SMTP_FROM = 'dev@lma.local';
    process.env.VAULT_ADDR = 'http://localhost:8200';
    process.env.VAULT_ROLE_ID = 'dev-role-id';
    process.env.VAULT_SECRET_ID = 'vaultsid-test';
    process.env.JWT_PUBLIC_KEY = PLACEHOLDER_KEY;
  };

  describe('loadConfig', () => {
    it('should load valid config from environment variables', () => {
      setValidEnv();

      const config = loadConfig();

      expect(config.env).toBe('dev');
      expect(config.service.name).toBe('notification-service');
      expect(config.service.port).toBe(7902);
      expect(config.observability.serviceName).toBe('notification-service');
      expect(config.observability.otlpEndpoint).toBe('http://localhost:4318');
      expect(config.nats.url).toBe('nats://localhost:4222');
      expect(config.redis.url).toBe('redis://localhost:6379');
      expect(config.postgres.host).toBe('localhost');
      expect(config.postgres.port).toBe(5432);
      expect(config.smtp.host).toBe('localhost');
      expect(config.smtp.port).toBe(1025);
      expect(config.vault.addr).toBe('http://localhost:8200');
      expect(config.auth.jwtPublicKey).toBe(PLACEHOLDER_KEY);
    });

    it('should throw on missing SERVICE_NAME', () => {
      setValidEnv();
      delete process.env.SERVICE_NAME;

      expect(() => loadConfig()).toThrow('Missing required environment variables: SERVICE_NAME');
    });

    it('should throw on missing SERVICE_PORT', () => {
      setValidEnv();
      delete process.env.SERVICE_PORT;

      expect(() => loadConfig()).toThrow('Missing required environment variables: SERVICE_PORT');
    });

    it('should apply default NATS_URL when not provided', () => {
      setValidEnv();
      delete process.env.NATS_URL;

      const config = loadConfig();

      expect(config.nats.url).toBe('nats://localhost:4222');
    });

    it('should parse SMTP config with all fields', () => {
      setValidEnv();
      process.env.SMTP_HOST = 'smtp.example.com';
      process.env.SMTP_PORT = '587';
      process.env.SMTP_USER = 'user@example.com';
      process.env.SMTP_PASSWORD = ['not', 'a', 'cred'].join('-');
      process.env.SMTP_FROM = 'noreply@example.com';

      const config = loadConfig();

      expect(config.smtp.host).toBe('smtp.example.com');
      expect(config.smtp.port).toBe(587);
      expect(config.smtp.user).toBe('user@example.com');
      expect(config.smtp.password).toBe(PLACEHOLDER_CRED);
      expect(config.smtp.from).toBe('noreply@example.com');
    });

    it('should throw on invalid SMTP_PORT', () => {
      setValidEnv();
      process.env.SMTP_PORT = 'invalid';

      expect(() => loadConfig()).toThrow('SMTP_PORT must be a valid number');
    });

    it('should parse Postgres config from DSN', () => {
      setValidEnv();
      delete process.env.PG_HOST;
      delete process.env.PG_PORT;
      delete process.env.PG_DATABASE;
      delete process.env.PG_USER;
      delete process.env.PG_PASSWORD;
      process.env.PG_DSN = DSN;

      const config = loadConfig();

      expect(config.postgres.host).toBe('pghost');
      expect(config.postgres.port).toBe(5433);
      expect(config.postgres.database).toBe('testdb');
      expect(config.postgres.user).toBe('user');
      expect(config.postgres.password).toBe('placeholder');
    });

    it('should parse Postgres config from individual variables', () => {
      setValidEnv();
      process.env.PG_HOST = 'db.example.com';
      process.env.PG_PORT = '5433';
      process.env.PG_DATABASE = 'mydb';
      process.env.PG_USER = 'dbuser';
      process.env.PG_PASSWORD = ['not', 'a', 'cred'].join('-');

      const config = loadConfig();

      expect(config.postgres.host).toBe('db.example.com');
      expect(config.postgres.port).toBe(5433);
      expect(config.postgres.database).toBe('mydb');
      expect(config.postgres.user).toBe('dbuser');
      expect(config.postgres.password).toBe(PLACEHOLDER_CRED);
    });

    it('should parse Vault config for AppRole auth', () => {
      setValidEnv();
      process.env.VAULT_ADDR = 'https://vault.example.com';
      process.env.VAULT_ROLE_ID = 'example-role-id';
      process.env.VAULT_SECRET_ID = 'vaultsid-test';

      const config = loadConfig();

      expect(config.vault.addr).toBe('https://vault.example.com');
      expect(config.vault.roleId).toBe('example-role-id');
      expect(config.vault.secretId).toBe('vaultsid-test');
    });

    it('should mark optional providers as undefined when not set', () => {
      setValidEnv();
      delete process.env.RESEND_API_KEY;
      delete process.env.TWILIO_ACCOUNT_SID;
      delete process.env.SLACK_WEBHOOK_URL;

      const config = loadConfig();

      expect(config.channels.resendApiKey).toBeUndefined();
      expect(config.channels.twilioAccountSid).toBeUndefined();
      expect(config.channels.twilioAuthToken).toBeUndefined();
      expect(config.channels.slackWebhookUrl).toBeUndefined();
    });

    it('should include optional providers when environment variables set', () => {
      setValidEnv();
      process.env.RESEND_API_KEY = 'resend' + '-cred-' + 'test';
      process.env.TWILIO_ACCOUNT_SID = 'twilio' + '-sid-' + 'test';
      process.env.TWILIO_AUTH_TOKEN = 'twilio' + '-cred-' + 'test';
      process.env.SLACK_WEBHOOK_URL = PLACEHOLDER_URL;

      const config = loadConfig();

      expect(config.channels.resendApiKey).toBe('resend-cred-test');
      expect(config.channels.twilioAccountSid).toBe('twilio-sid-test');
      expect(config.channels.twilioAuthToken).toBe('twilio-cred-test');
      expect(config.channels.slackWebhookUrl).toBe(PLACEHOLDER_URL);
    });

    it('should provide defaults for templates and queue config', () => {
      setValidEnv();
      delete process.env.TEMPLATES_DIR;
      delete process.env.QUEUE_BATCH_SIZE;
      delete process.env.DISPATCH_TICK_MS;
      delete process.env.DEFAULT_MAX_ATTEMPTS;
      delete process.env.NOTIFICATIONS_SUBJECT;

      const cfg = loadConfig();

      expect(cfg.templates.dir).toBe('src/templates');
      expect(cfg.queue.batchSize).toBe(10);
      expect(cfg.queue.tickMs).toBe(1000);
      expect(cfg.queue.defaultMaxAttempts).toBe(3);
      expect(cfg.nats.subjects.notifications).toBe('snapshots');
    });

    it('should parse templates and queue overrides from env', () => {
      setValidEnv();
      process.env.TEMPLATES_DIR = 'templates';
      process.env.QUEUE_BATCH_SIZE = '25';
      process.env.DISPATCH_TICK_MS = '500';
      process.env.DEFAULT_MAX_ATTEMPTS = '5';
      process.env.NOTIFICATIONS_SUBJECT = 'events.snapshots';

      const cfg = loadConfig();

      expect(cfg.templates.dir).toBe('templates');
      expect(cfg.queue.batchSize).toBe(25);
      expect(cfg.queue.tickMs).toBe(500);
      expect(cfg.queue.defaultMaxAttempts).toBe(5);
      expect(cfg.nats.subjects.notifications).toBe('events.snapshots');
    });
  });
});
